File: src/csharp/4_BehavioralPatterns/4_Iterator/ArrayBackedProperties.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using System.Text;
using System.Windows;
using ImpromptuInterface;
using JetBrains.dotMemoryUnit;
using NUnit.Framework;
using static System.Console;

namespace DesignPatterns
{
  public class Creature : IEnumerable<int>
  {
    private int [] stats = new int[3];

    private const int strength = 0;

    public int Strength
    {
      get => stats[strength];
      set => stats[strength] = value;
    }

    public int Agility { get; set; }
    public int Intelligence { get; set; }

    public double AverageStat => 
      stats.Average();

    public IEnumerator<int> GetEnumerator()
    {
      return stats.AsEnumerable().GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
      return GetEnumerator();
    }

    public int this[int index]
    {
      get { return stats[index]; }
      set { stats[index] = value; }
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
    }
  }
}

File: src/swift/4_BehavioralPatterns/4_Iterator/ArrayBackedProperties.swift
import Foundation
import XCTest

class Creature : Sequence
{
  var stats = [Int](repeating: 0, count: 3)

  private let _strength = 0
  private let _agility = 1
  private let _intelligence = 2

  // start with var strength: Int = 0
  var strength: Int
  {
    get { return stats[_strength] }
    set(value) { stats[_strength] = value }
  }

  var agility: Int
  {
    get { return stats[_agility] }
    set(value) { stats[_agility] = value }
  }

  var intelligence: Int
  {
    get { return stats[_intelligence] }
    set(value) { stats[_intelligence] = value }
  }

  var averageStat: Int
  {
    return stats.reduce(0, +) / stats.count
  }

  subscript(index: Int) -> Int
  {
    get { return stats[index] }
    set(value) { stats[index] = value }
  }

  func makeIterator() -> IndexingIterator<Array<Int>>
  {
    return IndexingIterator(_elements: stats)
  }
}

func main()
{
  let c = Creature()
  c.strength = 10
  c.agility = 15
  c.intelligence = 11

  print("strength = \(c.strength)")
  print("average stat = \(c.averageStat)")

  for s in c // no need for 'in c.stats'
  {
    print(s)
  }
}

main()